### ISSUE TITLE
Problem with pretrained model on MIMIC-CXR or MIMIC-CXR-JPG

### ISSUE THREAD

**Issue by LiamLiu62** [mlmed/torchxrayvision#164]
Hi! Thank you so much for your amazing works.

I want to know the pretrained model "densenet121-res224-all" was trained on MIMIC-CXR (DICOM version) or MIMIC-CXR-JPG. I find that my results on my MIMIC-CXR-JPG test set are worse than the Benchmark shown.

Thanks.

**Comment by ieee8023** [comment]
It was trained on the JPG files. To confirm I verified the dataset does not use a dicom loader to read the files: https://github.com/mlmed/torchxrayvision/blob/73effd43b1f664ade43bb5043c8c27422663f89a/torchxrayvision/datasets.py#L1219

The script for generating the benchmark results is here: https://github.com/mlmed/torchxrayvision/blob/master/scripts/model_calibrate.py
Maybe there is some difference you can see?

**Comment by LiamLiu62** [comment]
`import os
import pandas as pd
import numpy as np
import torch
import torchxrayvision as xrv
from PIL import Image
from sklearn.metrics import roc_auc_score
import torchvision.transforms as TF
from tqdm import tqdm

# Define target classes we want to evaluate with their model mappings
CLASS_MAPPINGS = {
    'Atelectasis': 'Atelectasis',
    'Cardiomegaly': 'Cardiomegaly', 
    'Consolidation': 'Consolidation',
    'Edema': 'Edema',
    'Enlarged Cardiomediastinum': 'Enlarged Cardiomediastinum',
    'Fracture': 'Fracture',
    'Lung Lesion': 'Lung Lesion',
    'Lung Opacity': 'Lung Opacity',
    'Pleural Effusion': 'Effusion',
    'Pneumonia': 'Pneumonia',
    'Pneumothorax': 'Pneumothorax'
}

TARGET_CLASSES = list(CLASS_MAPPINGS.keys())
MODEL_CLASSES = list(CLASS_MAPPINGS.values())

def load_image(image_path):
    """
    Load and preprocess a single chest X-ray image
    
    Args:
        image_path (str): Path to the image file
        
    Returns:
        torch.Tensor: Preprocessed image tensor ready for model input
    """

    # Load image with PIL
    img = Image.open(image_path).convert('L')  # Convert to grayscale
    img = np.array(img)
        
    # Normalize and process
    img = xrv.datasets.normalize(img, 255)
    img = img[None, ...]  # Add channel dimension
        
    transform = TF.Compose([
        xrv.datasets.XRayCenterCrop(),
        xrv.datasets.XRayResizer(224)
    ])
            
    img = transform(img)
    return torch.from_numpy(img)
        

def get_predictions(model, image_paths, device):
    """
    Get model predictions for a list of images
    
    Args:
        model: TorchXRayVision model
        image_paths (List[str]): List of image file paths
        device: PyTorch device
        
    Returns:
        np.ndarray: Predictions array of shape (n_samples, n_classes)
    """
    predictions = []
    
    for path in tqdm(image_paths, desc="Getting predictions"):
        img = load_image(path)
        if img is None:
            continue
            
        with torch.no_grad():
            img = img.to(device)
            output = model(img[None,...])
            predictions.append(output[0].cpu().numpy())
            
    return np.array(predictions)

def calculate_auroc(gt_labels, predictions, class_names):
    """
    Calculate per-class and overall AUROC scores
    
    Args:
        gt_labels (np.ndarray): Ground truth labels
        predictions (np.ndarray): Model predictions
        class_names (List[str]): Names of the classes
        
    Returns:
        dict: Dictionary containing per-class and overall AUROC scores
    """
    results = {}
    
    # Calculate per-class AUROC
    for i, class_name in enumerate(class_names):
        try:
            score = roc_auc_score(gt_labels[:, i], predictions[:, i])
            results[class_name] = score
        except ValueError:
            print(f"Could not calculate AUROC for {class_name}")
            results[class_name] = None
            
    # Calculate micro and macro AUROC
    valid_scores = [score for score in results.values() if score is not None]
    results['Macro AUROC'] = np.mean(valid_scores)
    
    try:
        results['Micro AUROC'] = roc_auc_score(
            gt_labels.ravel(), 
            predictions.ravel()
        )
    except ValueError:
        results['Micro AUROC'] = None
        
    return results

def main():
    """
    Main function to run the evaluation
    """
    # Set device
    device = torch.device("mps")
    
    # Load model
    model = xrv.models.DenseNet(weights="densenet121-res224-all")
    model = model.to(device)
    model.eval()
    
    # Read test CSV
    df = pd.read_csv('datasets/MIMIC_CXR/final/test.csv')
    
    # Process image paths - get image names and construct full paths
    df['full_path'] = df['image_path'].apply(lambda x: os.path.join('results/test_gt', x.split('/')[-1]))
    
    # Get model predictions
    predictions = get_predictions(model, df['full_path'].tolist(), device)
    
    # Get model's class names and map to indices for our target classes
    model_classes = model.pathologies
    class_indices = [model_classes.index(MODEL_CLASSES[i]) for i in range(len(TARGET_CLASSES))]
    predictions = predictions[:, class_indices]
    
    # Process ground truth labels
    gt_labels = df[TARGET_CLASSES].fillna(0).values
    gt_labels = (gt_labels == 1.0).astype(float)
    
    # Calculate AUROC scores
    results = calculate_auroc(gt_labels, predictions, TARGET_CLASSES)
    
    # Print results
    print("\nAUROC Scores:")
    print("-" * 40)
    for class_name, score in results.items():
        if score is not None:
            print(f"{class_name}: {score:.4f}")
        else:
            print(f"{class_name}: N/A")

if __name__ == "__main__":
    main() `


AUROC Scores:
----------------------------------------
Atelectasis: 0.6983
Cardiomegaly: 0.7338
Consolidation: 0.6643
Edema: 0.7971
Enlarged Cardiomediastinum: 0.6224
Fracture: 0.4939
Lung Lesion: 0.5997
Lung Opacity: 0.6412
Pleural Effusion: 0.8281
Pneumonia: 0.5976
Pneumothorax: 0.6264
Macro AUROC: 0.6639
Micro AUROC: 0.7822

I didn’t find any mistakes. Can you help me solve the problem?

**Comment by LiamLiu62** [comment]
Here is my code, can you help me find where I'm wrong? :)

import os
import pandas as pd
import numpy as np
import torch
import torchxrayvision as xrv
from PIL import Image
from sklearn.metrics import roc_auc_score
import torchvision.transforms as TF
from tqdm import tqdm

# Define target classes we want to evaluate with their model mappings
CLASS_MAPPINGS = {
    'Atelectasis': 'Atelectasis',
    'Cardiomegaly': 'Cardiomegaly', 
    'Consolidation': 'Consolidation',
    'Edema': 'Edema',
    'Enlarged Cardiomediastinum': 'Enlarged Cardiomediastinum',
    'Fracture': 'Fracture',
    'Lung Lesion': 'Lung Lesion',
    'Lung Opacity': 'Lung Opacity',
    'Pleural Effusion': 'Effusion',
    'Pneumonia': 'Pneumonia',
    'Pneumothorax': 'Pneumothorax'
}

TARGET_CLASSES = list(CLASS_MAPPINGS.keys())
MODEL_CLASSES = list(CLASS_MAPPINGS.values())

def load_image(image_path):
    """
    Load and preprocess a single chest X-ray image
    
    Args:
        image_path (str): Path to the image file
        
    Returns:
        torch.Tensor: Preprocessed image tensor ready for model input
    """

    # Load image with PIL
    img = Image.open(image_path).convert('L')  # Convert to grayscale
    img = np.array(img)
        
    # Normalize and process
    img = xrv.datasets.normalize(img, 255)
    img = img[None, ...]  # Add channel dimension
        
    transform = TF.Compose([
        xrv.datasets.XRayCenterCrop(),
        xrv.datasets.XRayResizer(224)
    ])
            
    img = transform(img)
    return torch.from_numpy(img)
        

def get_predictions(model, image_paths, device):
    """
    Get model predictions for a list of images
    
    Args:
        model: TorchXRayVision model
        image_paths (List[str]): List of image file paths
        device: PyTorch device
        
    Returns:
        np.ndarray: Predictions array of shape (n_samples, n_classes)
    """
    predictions = []
    
    for path in tqdm(image_paths, desc="Getting predictions"):
        img = load_image(path)
        if img is None:
            continue
            
        with torch.no_grad():
            img = img.to(device)
            output = model(img[None,...])
            predictions.append(output[0].cpu().numpy())
            
    return np.array(predictions)

def calculate_auroc(gt_labels, predictions, class_names):
    """
    Calculate per-class and overall AUROC scores
    
    Args:
        gt_labels (np.ndarray): Ground truth labels
        predictions (np.ndarray): Model predictions
        class_names (List[str]): Names of the classes
        
    Returns:
        dict: Dictionary containing per-class and overall AUROC scores
    """
    results = {}
    
    # Calculate per-class AUROC
    for i, class_name in enumerate(class_names):
        try:
            score = roc_auc_score(gt_labels[:, i], predictions[:, i])
            results[class_name] = score
        except ValueError:
            print(f"Could not calculate AUROC for {class_name}")
            results[class_name] = None
            
    # Calculate micro and macro AUROC
    valid_scores = [score for score in results.values() if score is not None]
    results['Macro AUROC'] = np.mean(valid_scores)
    
    try:
        results['Micro AUROC'] = roc_auc_score(
            gt_labels.ravel(), 
            predictions.ravel()
        )
    except ValueError:
        results['Micro AUROC'] = None
        
    return results

def main():
    """
    Main function to run the evaluation
    """
    # Set device
    device = torch.device("mps")
    
    # Load model
    model = xrv.models.DenseNet(weights="densenet121-res224-all")
    model = model.to(device)
    model.eval()
    
    # Read test CSV
    df = pd.read_csv('datasets/MIMIC_CXR/final/test.csv')
    
    # Process image paths - get image names and construct full paths
    df['full_path'] = df['image_path'].apply(lambda x: os.path.join('results/test_gt', x.split('/')[-1]))
    
    # Get model predictions
    predictions = get_predictions(model, df['full_path'].tolist(), device)
    
    # Get model's class names and map to indices for our target classes
    model_classes = model.pathologies
    class_indices = [model_classes.index(MODEL_CLASSES[i]) for i in range(len(TARGET_CLASSES))]
    predictions = predictions[:, class_indices]
    
    # Process ground truth labels
    gt_labels = df[TARGET_CLASSES].fillna(0).values
    gt_labels = (gt_labels == 1.0).astype(float)
    
    # Calculate AUROC scores
    results = calculate_auroc(gt_labels, predictions, TARGET_CLASSES)
    
    # Print results
    print("\nAUROC Scores:")
    print("-" * 40)
    for class_name, score in results.items():
        if score is not None:
            print(f"{class_name}: {score:.4f}")
        else:
            print(f"{class_name}: N/A")

if __name__ == "__main__":
    main()
